// @proton/vert/dist/vert.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.log = exports.Vert = void 0;
const memory_1 = require("./memory");
const loglevel_1 = __importDefault(require("loglevel"));
const loglevel_plugin_prefix_1 = __importDefault(require("loglevel-plugin-prefix"));
let log = loglevel_1.default;
exports.log = log;
loglevel_plugin_prefix_1.default.reg(log);
loglevel_plugin_prefix_1.default.apply(log);
try {
    log.setLevel(process.env.LOG_LEVEL || 'warn');
}
catch (e) {
}
class Vert {
    get memory() {
        return this._memory;
    }
    constructor(imports, bytes) {
        const getReady = async () => {
            bytes = await Promise.resolve(bytes);
            const { module, instance } = await WebAssembly.instantiate(bytes, imports);
            this.module = module;
            this.instance = instance;
            this._memory = new memory_1.Memory(this.instance.exports.memory);
        };
        this.ready = getReady();
    }
}
exports.Vert = Vert;
(function (Vert) {
    function setLogger(logger) {
        exports.log = log = logger;
    }
})(Vert || (Vert = {}));
exports.Vert = Vert;
