// @proton/vert/dist/proton/bn.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.symbolCodeToBigInt = exports.nameTypeToBigInt = exports.bigIntToName = exports.nameToBigInt = exports.bigIntToBn = exports.bnToBigInt = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
const eosio_1 = require("@greymass/eosio");
function bnToBigInt(num) {
    return BigInt(num.toString());
}
exports.bnToBigInt = bnToBigInt;
function bigIntToBn(num) {
    return new bn_js_1.default(num.toString());
}
exports.bigIntToBn = bigIntToBn;
function nameToBigInt(name) {
    return bnToBigInt(eosio_1.Name.from(name).value.value);
}
exports.nameToBigInt = nameToBigInt;
function bigIntToName(name) {
    return eosio_1.Name.from(eosio_1.UInt64.from(bigIntToBn(name)));
}
exports.bigIntToName = bigIntToName;
function nameTypeToBigInt(nameType) {
    return nameToBigInt(eosio_1.Name.from(nameType));
}
exports.nameTypeToBigInt = nameTypeToBigInt;
function symbolCodeToBigInt(symbolCode) {
    return bnToBigInt(symbolCode.value.value);
}
exports.symbolCodeToBigInt = symbolCodeToBigInt;
