// @greymass/eosio/lib/modules/types.js
/**
 * EOSIO Core v0.6.8
 * https://github.com/greymass/eosio-core
 *
 * @license
 * Copyright (c) 2020 FFF00 Agents AB & Greymass Inc. All Rights Reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 * 
 *  1. Redistribution of source code must retain the above copyright notice, this
 *     list of conditions and the following disclaimer.
 * 
 *  2. Redistribution in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 * 
 *  3. Neither the name of the copyright holder nor the names of its contributors
 *     may be used to endorse or promote products derived from this software without
 *     specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * YOU ACKNOWLEDGE THAT THIS SOFTWARE IS NOT DESIGNED, LICENSED OR INTENDED FOR USE
 * IN THE DESIGN, CONSTRUCTION, OPERATION OR MAINTENANCE OF ANY MILITARY FACILITY.
 */
import 'tslib';
import './bytes.js';
export { a6 as AccountObject, _ as AccountPermission, a2 as AccountRefundRequest, $ as AccountResourceLimit, a5 as AccountRexInfo, a4 as AccountRexInfoMaturities, a1 as AccountSelfDelegatedBandwidth, a0 as AccountTotalResources, a3 as AccountVoterInfo, ag as ActiveSchedule, af as ActiveScheduleProducer, ae as ActiveScheduleProducerAuthority, a9 as BlockExtension, ah as BlockStateHeader, an as GetActionsResponse, ai as GetBlockHeaderStateResponse, ad as GetBlockResponse, ac as GetBlockResponseTransactionReceipt, at as GetControlledAccountsResponse, aj as GetInfoResponse, as as GetKeyAccountsResponse, al as GetTableByScopeResponse, ak as GetTableByScopeResponseRow, ar as GetTransactionResponse, au as GetTransactionStatusResponse, aa as HeaderExtension, a8 as NewProducers, a7 as NewProducersEntry, am as OrderedActionsResult, aq as TransactionInfo, ao as TransactionTrace, ap as Trx, ab as TrxVariant } from './client-5963ba48.js';
import './key-type.js';
import './integer.js';
import 'brorand';
import 'elliptic';
import './utils.js';
import './provider.js';
import 'hash.js';
import './type-alias.js';
import 'bn.js';
